src/ekat/util/ekat_meta_utils.hpp

```diff
@@ -0,0 +1,235 @@
+    l(T());
+  }
+};
+
+template<typename T, typename...Ts>
+struct TypeListFor<TypeList<T,Ts...>> {
+  template<typename Lambda>
+  constexpr TypeListFor (Lambda&& l) {
+    l(T());
+    TypeListFor<TypeList<Ts...>> tlf(l);
+  }
+};
+
+// A map of types
+template<typename KP, typename VP>
+struct TypeMap;
+
+template<typename... Ks, typename... Vs>
+struct TypeMap<TypeList<Ks...>,TypeList<Vs...>> : public std::tuple<Vs...> {
+  struct NotFound {};
+private:
+
+  template<typename T>
+  using base_t = typename std::remove_cv<typename std::remove_reference<T>::type>::type;
+
+  template<typename V1, typename V2>
+  static typename std::enable_if<std::is_same<base_t<V1>,base_t<V2>>::value,bool>::type
+  check_eq (const V1& v1, const V2& v2) {
+    return v1==v2;
+  }
+  template<typename V1, typename V2>
+  static typename std::enable_if<!std::is_same<base_t<V1>,base_t<V2>>::value,bool>::type
+  check_eq (const V1& /* v1 */, const V2& /* v2 */) {
+    return false;
+  }
+public:
+
+  using keys = TypeList<Ks...>;
+  using vals = TypeList<Vs...>;
+  using self = TypeMap<keys,vals>;
+  static_assert (keys::size==vals::size, "Error! Keys and values have different sizes.\n");
+  static_assert (UniqueTypeList<keys>::value, "Error! Keys are not unique.\n");
+
+  template<typename K>
+  struct AtHelper {
+    static constexpr int pos = FirstOf<K,keys>::pos;
+  private:
+    // The second arg to std::conditional must be instantiated. If pos<0,
+    // AccessList does not compile. Hence, if pos<0, use 0, which will for
+    // sure work (TypeList has size>=1). The corresponding type will not
+    // be exposed anyways. It's just to get a valid instantiation of
+    // std::conditional.
+    struct GetSafe {
+      static constexpr int safe_pos = pos>=0 ? pos : 0;
+      using type = typename AccessList<vals,safe_pos>::type;
+    };
+  public:
+    using type = typename std::conditional<(pos<0),NotFound,typename GetSafe::type>::type;
+  };
+
+  template<typename K>
+  static constexpr bool has_t () {
+    return not std::is_same<at_t<K>,NotFound>::value;
+  }
+
+  template<typename V>
+  bool has_v (const V& v) const {
+    return has_v_impl<(FirstOf<V,vals>::pos>=0)>(v,*this);
+  }
+
+
+  template<typename K>
+  using at_t = typename AtHelper<K>::type;
+
+  template<typename K>
+  at_t<K>& at () {
+    static_assert (has_t<K>(), "Error! Type not found in this TypeMap's keys.\n");
+    return std::get<AtHelper<K>::pos>(*this);
+  }
+
+  template<typename K>
+  const at_t<K>& at () const {
+    static_assert (has_t<K>(), "Error! Type not found in this TypeMap's keys.\n");
+    return std::get<AtHelper<K>::pos>(*this);
+  }
+
+  static constexpr int size = keys::size;
+private:
+
+  template<bool has_V_type, typename V>
+  typename std::enable_if<has_V_type,bool>::type
+  has_v_impl (const V& v, const self& map) const {
+    bool found = false;
+    TypeListFor<keys>([&](auto t){
+      using key_t = decltype(t);
+      const auto& value = map.at<key_t>();
```

Comment:
This line seems to be tripping up Mappy's C++ compiler: 

```
In file included from /ascldap/users/e3sm-jenkins/jenkins-ws/workspace/EKAT_PullRequest_Autotester_Mappy/213/ekat-src/tests/utils/meta_utils.cpp:3:
/ascldap/users/e3sm-jenkins/jenkins-ws/workspace/EKAT_PullRequest_Autotester_Mappy/213/ekat-src/src/ekat/util/ekat_meta_utils.hpp: In lambda function:
/ascldap/users/e3sm-jenkins/jenkins-ws/workspace/EKAT_PullRequest_Autotester_Mappy/213/ekat-src/src/ekat/util/ekat_meta_utils.hpp:217:39: error: expected primary-expression before > token
  217 |       const auto& value = map.at();
      |                                       ^
/ascldap/users/e3sm-jenkins/jenkins-ws/workspace/EKAT_PullRequest_Autotester_Mappy/213/ekat-src/src/ekat/util/ekat_meta_utils.hpp:217:41: error: expected primary-expression before ) token
  217 |       const auto& value = map.at();
```

Reply:
That said, I am not sure why line 217 reported in the error message doesn't include the template parameter for `at`.

Reply:
Really strange. I saw some (somewhat) similar errors while developing this branch, with the compiler reporting errors at seemingly unrelated lines, or with lines looking slightly different. I wonder if the heavy meta-programming is flipping out the compiler...

I'll manually build on mappy, and find the bug.